syntax = "proto3";

package bmp_grapher_proto;

message RGB {
  int32 red = 1;
  int32 green = 2;
  int32 blue = 3;
}

message SingleColorScheme {
  int32 scaling_denominator = 1;
  RGB hundred_per_cent_color = 2;
  RGB scaling_base_color = 3;
}

message DodgyColorScheme {
  RGB solid_color = 1;
}

message ColorConfig {
  string out_filename = 1;

  oneof color_scheme {
    SingleColorScheme single_color_scheme = 2;
    DodgyColorScheme dodgy_color_scheme = 3;
  }
}

message MandelbrotCalc {
  int32 max_iterations = 1;
}

message BmpGrapher {
  double x_center = 1;
  double y_center = 2;
  double x_domain_width = 3;
  int32 pixel_width = 4;
  int32 pixel_height = 5;

  // you must choose one and only one calculator per bitmap...
  oneof pixel_score_calculator {
    MandelbrotCalc mandelbrot_calc = 6;
  }
  
  // but you can generate as many re-colorizations as you like.
  repeated ColorConfig color_configs = 7;
}

message BmpGrapherFactory {
  repeated BmpGrapher bitmaps = 1;
}